File: Xit/XTTextPreviewController.m
#import "XTTextPreviewController.h"

#import <WebKit/WebKit.h>

#import "XTRepository+Parsing.h"
#import "Xit-Swift.h"

@implementation XTTextPreviewController

- (instancetype)initWithNibName:(NSString*)nibNameOrNil bundle:(NSBundle*)nibBundleOrNil
{
  self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
  if (self != nil) {
    // Initialization code here.
  }
  
  return self;
}

- (void)clear
{
  [_webView.mainFrame loadHTMLString:@"" baseURL:nil];
}

- (void)loadText:(NSString*)text
{
  if (text == nil)
    text = @"";

  NSMutableString *textLines = [NSMutableString string];

  [text enumerateLinesUsingBlock:^(NSString *line, BOOL *stop) {
    [textLines appendFormat:@"<div>%@</div>\n",
                            [[self class] escapeText:line]];
  }];

  NSString *htmlTemplate = [[self class] htmlTemplate:@"text"];
  NSString *html = [NSString stringWithFormat:htmlTemplate, textLines];

  [_webView.mainFrame loadHTMLString:html baseURL:[[self class] baseURL]];
}

- (void)loadData:(NSData*)data
{
  if (data == nil)
    return;

  NSString *text = [[NSString alloc]
                    initWithData:data encoding:NSUTF8StringEncoding];
  
  if (text == nil)
    text = [[NSString alloc]
            initWithData:data encoding:NSUTF16StringEncoding];
  
  [self loadText:text];
}

- (BOOL)isFileChanged:(NSString*)path inRepository:(XTRepository*)repo
{
  XTWindowController *controller = self.view.window.windowController;
  NSArray *changes =
      [repo changesForRef:controller.selectedModel.shaToSelect parent:nil];

  for (XTFileChange *change in changes)
    if ([change.path isEqualToString:path])
      return YES;
  return NO;
}

- (void)loadPath:(NSString *)path
           model:(id<XTFileChangesModel>)model
          staged:(BOOL)staged
{
  [self loadData:[model dataForFile:path staged:staged]];
}

- (void)loadUnstagedPath:(NSString*)path
              repository:(XTRepository*)repository
{
  NSURL *url = [repository.repoURL URLByAppendingPathComponent:path];
  
  [self loadData:[NSData dataWithContentsOfURL:url]];
}

- (void)loadStagedPath:(NSString*)path
            repository:(XTRepository*)repository
{
  NSError *error = nil;
  
  [self loadData:[repository contentsOfStagedFile:path error:&error]];
}

@end

File: Xit/XTTextPreviewController.swift
import Foundation

class XTTextPreviewController: WebViewController
{
  var isLoaded: Bool = false
  
  override func wrappingWidthAdjustment() -> Int
  {
    return 6
  }

  func minString(_ string: String) -> String
  {
    return string.isEmpty ? "\n" : string
  }
  
  func load(text: String?)
  {
    let text = text ?? ""
    let lines = text.components(separatedBy: .newlines).map {
      "<div>\(minString($0.xmlEscaped))</div>"
    }
    let textLines = lines.joined(separator: "\n")
    let htmlTemplate = WebViewController.htmlTemplate("text")
    let html = String(format: htmlTemplate, textLines)
    
    webView?.mainFrame.loadHTMLString(html,
                                      baseURL: XTTextPreviewController.baseURL)
    isLoaded = true
  }
  
  func load(data: Data?)
  {
    guard let data = data,
          let text = String(data: data, encoding: .utf8) ??
                     String(data: data, encoding: .utf16)
    else { return }
    
    load(text: text)
  }
}

extension XTTextPreviewController: XTFileContentController
{
  public func clear()
  {
    webView?.mainFrame.loadHTMLString("", baseURL: nil)
    isLoaded = false
  }
  
  public func load(path: String!, selection: RepositorySelection!, staged: Bool)
  {
    load(data: selection.list(staged: staged).dataForFile(path))
  }
}
